### ISSUE TITLE
Checking for conflicts before registering synthetic child providers/filters doesn't work correctly with regex

### ISSUE THREAD

**Issue by slackito** [llvm/llvm-project#57947]
`CommandObjectTypeSynthAdd::AddSynth` has some code to check for conflicting filters in the same category (there's a symmetric check for conflicting synth providers when adding a new filter, but let's focus on only one case). The check looks like this:

```c++
  if (category->AnyMatches(type_name, eFormatCategoryItemFilter, false)) {
    if (error)
      error->SetErrorStringWithFormat("cannot add synthetic for type %s when "
                                      "filter is defined in same category!",
                                      type_name.AsCString());
    return false;
  }
```

The important detail here is that `type_name` can actually be either a type name **or a regex** (if using `type synthetic add -x`). However, `category->AnyMatches` always treats `type_name` as a type name, and will try to match a regex string against other regex strings, which is not correct.

For example, this is lldb working correctly and rejecting a conflicting formatter.
```
(lldb) script
Python Interactive Interpreter. To exit, type 'quit()', 'exit()' or Ctrl-D.
>>> class my_child_provider:
...   pass
... 
>>> ^D
now exiting InteractiveConsole...
(lldb) type synthetic add -l my_child_provider whatever
(lldb) type filter add --child a whatever
error: cannot add filter for type whatever when synthetic is defined in same category!
```

However, if we use regex it will happily accept conflicting formatters:
```
(lldb) type synthetic add -l my_child_provider -x "^whatever$"
(lldb) type filter add --child a -x "^whatever$"
(lldb) 
```
because the string "^whatever$" is not matched by the regex "^whatever$".

We can also make it fail the other way around, and make it incorrectly reject non-conflicting formatters. For example:
```
(lldb) type synthetic add -l my_child_provider -x "^MyType\[[0-9]+]$"
(lldb) type filter add --child a -x "MyType[12]"
error: cannot add filter for type MyType[12] when synthetic is defined in same category!
```
In this case, `"^MyType\[[0-9]+]$"` would accept types that look like an array of `MyType`, and `"MyType[12]"` _looks_ like an array, but it would actually match `MyType1` and `MyType2` so there's no conflict.

I think the right thing to do is to just skip this check when adding a regex-based filter. I'll prepare a patch.

**Comment by slackito** [comment]
Sent https://reviews.llvm.org/D134570 to skip the check if registering a new filter/synth using a regex.

**Comment by slackito** [comment]
Fixed in 69c661a65ff8c04ce9408a788a77df67481545a1 by restricting the check to "exact match" formatters only.
